log/sockstatlog/logger.go

```diff
@@ -0,0 +1,124 @@
+package sockstatlog
+
+import (
+	"encoding/json"
+	"path/filepath"
+	"time"
+
+	"tailscale.com/logtail/filch"
+	"tailscale.com/net/sockstats"
+)
+
+// pollPeriod specifies how often to poll for socket stats.
+const pollPeriod = time.Second / 10
+
+// Logger logs statistics about network sockets.
+type Logger struct {
+	ticker time.Ticker
+
+	// last is the set of socket stats we saw.
+	last     statMap
+	lastTime time.Time
+
+	logbuffer *filch.Filch
+}
+
+// statMap contains a map of stat values for labels.
+// In the net/sockstats package, a SockStat always represents total bytes transferred.
+// In this package, we reuse the type to also represent bytes transferred within a specific time interval.
+type statMap map[sockstats.Label]sockstats.SockStat
+
+type logEntry struct {
+	// Start and End of the time interval as Unix timestamps in milliseconds.
+	Start int64 `json:"start"`
+	End   int64 `json:"end"`
+
+	// Stats records the stats for the time interval.
+	Stats statMap `json:"stats"`
+}
+
+// NewLogger returns a new Logger that will store stats in logdir.
+// On platforms that do not support sockstat logging, a nil Logger will be returned.
+// The returned Logger is not yet running.
+func NewLogger(logdir string) (*Logger, error) {
+	if !loggingEnabled {
+		return nil, nil
+	}
+	filchPrefix := filepath.Join(logdir, "sockstats")
+	filch, err := filch.New(filchPrefix, filch.Options{ReplaceStderr: false})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Logger{
+		ticker:    *time.NewTicker(pollPeriod),
+		logbuffer: filch,
+	}, nil
+}
+
+func (l *Logger) Start() {
+	go l.poll()
+}
+
+// poll fetches the current socket stats at the configured time interval,
+// calculates the delta since the last poll, and logs any non-zero values.
+// This method does not return.
+func (l *Logger) poll() {
+	enc := json.NewEncoder(l.logbuffer)
+	for t := range l.ticker.C {
+		stats := sockstats.Get()
+		diffstats, changed := delta(l.last, stats.Stats)
+		if changed {
+			entry := logEntry{
+				Start: l.lastTime.UnixMilli(),
+				End:   t.UnixMilli(),
+				Stats: diffstats,
+			}
+			enc.Encode(entry)
+		}
+		l.lastTime = t
+		l.last = stats.Stats
+	}
+}
+
+func (l *Logger) Shutdown() {
+	l.ticker.Stop()
+	l.logbuffer.Close()
+}
+
+// delta calculates the delta stats between two snapshots.
+// a and b are maps of SockStat values containing total bytes transferred.
+// The returned map contains the bytes transferred between the two snapshots.
+// Zero values are omitted from the returned map, and changed indicates if any stats have changed during the interval.
+func delta(a, b statMap) (diffstats statMap, changed bool) {
+	diffstats = make(statMap, len(b))
+	for label, bs := range b {
+		var labelChanged bool
+		as := a[label]
+		stat := sockstats.SockStat{
+			TxBytes:            bs.TxBytes - as.TxBytes,
+			RxBytes:            bs.RxBytes - as.RxBytes,
+			TxBytesByInterface: make(map[string]int64, len(bs.TxBytesByInterface)),
```

Comment:
Storing the per-interface information seems like it'll be pretty bulky. Could we instead return the current interface in sockstats.SockStat, and include it in the delta if it changes? We can assume that all transfers in a delta were on the same interface, which is basically true at the 10hz sampling rate.

Reply:
This is particularly bad when we have multiple labels logged in a given time period, since each label repeats the interface.  However, I worry about dropping the interface, since it means you can't necessarily pick a random entry and know what interface it was on.  What about just logging it once per entry?

```
{"t":1678419626985,"d":99,"i":"en0","s":{"0":{"tx":55,"rx":710}}}
{"t":1678419627185,"d":100,"i":"en0","s":{"1":{"tx":332}}}
{"t":1678419627285,"d":99,"i":"en0","s":{"1":{"tx":1170,"rx":364}}}
{"t":1678419627484,"d":100,"i":"en0","s":{"1":{"tx":876,"rx":721}}}
{"t":1678419627685,"d":100,"i":"en0","s":{"1":{"tx":45,"rx":12453}}}
{"t":1678419627885,"d":99,"i":"en0","s":{"1":{"tx":1032,"rx":1018},"6":{"tx":113},"7":{"tx":365},"8":{"tx":960,"rx":352}}}
{"t":1678419627984,"d":99,"i":"en0","s":{"1":{"rx":58},"8":{"tx":600,"rx":528}}}
{"t":1678419628084,"d":99,"i":"en0","s":{"1":{"tx":1670},"2":{"tx":268},"6":{"tx":73},"7":{"tx":351},"8":{"tx":320,"rx":704}}}
```

This new update I'm about to push is also very close to allowing us to start coalescing some log records into short events which will save us on storage.  We had talked about doing that event aggregation only on the c2n call, but we could also do smaller interval joining for storage.  So for example, we could still grab stats at 10hz, write to disk in 300ms sessions, then respond to c2n at a longer control-defined session interval.  That's not what this code is doing yet, but I mention it here because it makes me feel a little more comfortable including the interface in every log entry.